beamer/agent/chain.py

```
@@ -625,83 +624,92 @@ def maybe_prove(claim: Claim, context: Context) -> None:
 
     future.add_done_callback(on_future_done)
-    context.l1_resolutions[proof_tx] = future
+    context.l1_resolutions[proved_tx] = future
 
-    context.logger.info("Initiated Optimism prove", request=request, claim=claim, tx_hash=proof_tx)
+    context.logger.info(
+        "Initiated Optimism prove", request=request, claim=claim, tx_hash=proved_tx
+    )
 
 
 def maybe_resolve(claim: Claim, context: Context) -> bool:
     request = context.requests.get(claim.request_id)
 
     assert request is not None, "Active claim for non-existent request"
 
-    if not claim.message_proved:
+    if claim.proved_tx is None or claim.proved_tx in context.l1_resolutions:
         return False
 
     if not _proof_ready_for_l1_relay(request) and not _invalidation_ready_for_l1_relay(claim):
         return False
 
     timestamp_finalized = False
 
-    if request.fill_timestamp:
-        timestamp_finalized = _timestamp_is_l1_finalized(
-            request.fill_timestamp, context, request.target_chain_id
-        )
-        assert request.fill_tx is not None
-        proof_tx = request.fill_tx
-    elif claim.invalidation_timestamp:
-        timestamp_finalized = _timestamp_is_l1_finalized(
-            claim.invalidation_timestamp, context, request.target_chain_id
-        )
-        assert claim.invalidation_tx is not None
-        proof_tx = claim.invalidation_tx
+    match claim.proved_tx:
+        case request.fill_tx:
+            assert request.fill_timestamp is not None
+            timestamp_finalized = _timestamp_is_l1_finalized(
+                request.fill_timestamp, context, request.target_chain_id
+            )
+        case claim.invalidation_tx:
+            assert claim.invalidation_timestamp is not None
+            timestamp_finalized = _timestamp_is_l1_finalized(
+                claim.invalidation_timestamp, context, request.target_chain_id
+            )
+        case _:
+            return False
```

Actually, I would raise an exception here, like I commented before, and not return `False`. Because if this happens, something seriously went wrong and we want to know about it.